# Logistic Regression Classifier

In [1]:
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score


In [2]:
# read the datasets 
train = pd.read_csv(r"D:/NEU/Fall 2020/NLP/Project/train.csv")
test = pd.read_csv(r"D:/NEU/Fall 2020/NLP/Project/test.csv")
submission_binary = pd.read_csv(r"D:/NEU/Fall 2020/NLP/Project/sample_submission.csv")

In [3]:
# vectorize data using tf_idf vectorizer
# for unigram vectorization

Vectorize = TfidfVectorizer(max_df = 0.8, ngram_range = (1,1))
X = Vectorize.fit_transform(train["comment_text"])

# binarize the target label
# positive class is 'toxic' class
y = np.where(train['target'] >= 0.5, 1, 0)
test_X = Vectorize.transform(test["comment_text"])


In [4]:
# split training set into 80% training and 20% validation sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)


In [5]:
# logistic regression model 

log = LogisticRegression(solver='sag', max_iter=1000, n_jobs=-1)
log.fit(X_train, y_train)
y_pred = log.predict(X_test)

In [6]:
# accuracy score of the model
print(accuracy_score(y_test, y_pred))


0.9474174111780594


**accuracy of the logistic regression model is 94.7%**

In [8]:
# confusion martix
from sklearn import metrics

print('Confusion matrix of the classifier')
print(metrics.confusion_matrix(y_test, y_pred, labels=[0,1]))

print('Advanced metrics')
print(metrics.classification_report(y_test, y_pred, labels=[0,1]))



Confusion matrix of the classifier
[[328612   3545]
 [ 15436  13382]]
Advanced metrics
              precision    recall  f1-score   support

           0       0.96      0.99      0.97    332157
           1       0.79      0.46      0.59     28818

    accuracy                           0.95    360975
   macro avg       0.87      0.73      0.78    360975
weighted avg       0.94      0.95      0.94    360975



In [13]:
import collections, numpy

collections.Counter(y_test)


Counter({0: 332157, 1: 28818})

There are 28818 toxic comments and 332157 non-toxic comments

**The precision for toxic class is 0.79 whereas the recall for positive class is 0.46 which implies that more than half 
the toxic comments have been labeled as non-toxic which is undesirable**